roachprod: fix node environment variable propagation

erikgrinaker opened the pull request:
In #67363, configured environment variables were no longer
propagated to nodes. In particular, the default node envvar
`COCKROACH_ENABLE_RPC_COMPRESSION=false` was not passed,
which caused a 5% performance regression on certain benchmarks.

This patch fixes propagation of environment variables.

Resolves #68887.

Release note: None

/cc @cockroachdb/test-eng 

---

I'm going on vacation next week, I'd appreciate it if you could get this merged for me.

erikgrinaker committed:
roachprod: fix node environment variable propagation

In #67363, configured environment variables were no longer
propagated to nodes. In particular, the default node envvar
`COCKROACH_ENABLE_RPC_COMPRESSION=false` was not passed,
which caused a 5% performance regression on certain benchmarks.

This patch fixes propagation of environment variables.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68960)
<!-- Reviewable:end -->

stevendanna commented on pkg/cmd/roachprod/install/cockroach.go:
```diff
@@ -436,10 +436,10 @@ func (h *crdbInstallHelper) generateStartCmd(
 			"GOTRACEBACK=crash",
 			"COCKROACH_SKIP_ENABLING_DIAGNOSTIC_REPORTING=1",
-		}, h.getEnvVars()...),
+		}, h.c.Env...), h.getEnvVars()...),
```
Just for reference, it looks like this used to happen in getEnvVars: https://github.com/cockroachdb/cockroach/pull/67363/files#diff-052e0342a18522387861c55bc7df04d8b6a888977f4593e2c8cd7c5749f94068L757-L761

stevendanna reviewed (approved):
Thanks for hunting this down.

stevendanna reviewed (dismissed):
It looks like we might be generating incorrect roachprod commands:

```
  cluster.go:1798,kv.go:630,test_runner.go:777: /go/src/github.com/cockroachdb/cockroach/bin/roachprod start --env COCKROACH_MEMPROF_INTERVAL=1m COCKROACH_DISABLE_QUIESCENCE=false --args=--cache=256MiB local:1-3 returned: exit status 1
```

The argument to `--env` needs to be comma-separated I believe, although I'm not sure why that would have changed here so perhaps I need to read this log more carefully :D.

stevendanna commented:
I don't mind picking this up while you are on PTO

stevendanna committed:
roachtest: pass env vars to roachprod via multiple --env flags

Roachprod now shell escapes environment variables when rendering its
start script. As a result, multiple environment variables passed as a
single string, such as:

    --env "COCKROACH_MEMPROF_INTERVAL=1m COCKROACH_DISABLE_QUIESCENCE=true"

would result in us exporting it as a single entry:

    export 'COCKROACH_MEMPROF_INTERVAL=1m COCKROACH_DISABLE_QUIESCENCE=true'

which results in a single environment variable getting exported

    COCKROACH_MEMPROF_INTERVAL='1m COCKROACH_DISABLE_QUIESCENCE=true'

I've converted the tests that were passing multiple environment
variables in this was to passing them via multiple invocations of the
`--env` flag.

This seem preferable than teaching roachprod more sophisticated
environment variable parsing.

Release note: None

stevendanna commented:
Found the cause of the test failure -- details in the commit message. This should be good for another look.

tbg reviewed (approved):
Thanks!

stevendanna commented:
bors r=tbg

erikgrinaker commented:
>Found the cause of the test failure -- details in the commit message. This should be good for another look.

Nice catch! Thanks for taking this across the line @stevendanna, appreciate it!
